### DISCUSSION TITLE
TypeScript type for `setQueryData` updater function

### DISCUSSION THREAD

**Discussion by ryands17** [TanStack/query#925] [General] [answered]
In `react-query` version `2.12.1`, the typescript types changed for `setQueryData` and it doesn't accept an undefined value anymore.

For context, this is to get the query data:

```ts
const previousTodos = queryCache.getQueryData<Todos>('todos') // returns Todos | undefined
```

Here, `setQueryData` doesn't accept `undefined` as the returned value:

```ts
queryCache.setQueryData<Todos>('todos', todos => previousTodos)
```

This gives the following TS issue:
![image](https://user-images.githubusercontent.com/19697099/91213247-ef72a000-e72e-11ea-9aac-7b88c8e503dd.png)

Was this change intentional in the newer version?

This is only an issue with the updater function and not with a direct update. Doing this works fine:

```ts
queryCache.setQueryData('todos', previousTodos)
```

**Comment by boschni** [comment, marked as answer]
Hi @ryands17! When specifying `setQueryData<Todos>` you are basically saying the result should be of type `Todos`. Because `setQueryData` sets the query status to success, and success means the data is there, this would be correct. You could specify the result as `setQueryData<Todos | undefined>` to also allow `undefined`.
